include/exadg/convection_diffusion/time_integration/time_int_bdf.cpp

```diff
@@ -591,6 +654,13 @@ TimeIntBDF<dim, Number>::set_velocities_and_times(
   times      = times_in;
 }
 
+template<int dim, typename Number>
+dealii::LinearAlgebra::distributed::Vector<Number> const &
+TimeIntBDF<dim, Number>::get_solution_np() const
+{
+  return (this->solution[0]);
```

Comment:
This class also has a member variable `solution_np`, right? Then, the implementation of this function seems to be ambiguous/undefined. Strictly speaking, `get_solution_np()` has to return `solution_np` in my opinion. Is the reason behind the place in the time stepping / AMR algorithm at which this function is called?

Generally speaking, the consistency of `time_step_number`, `time`, and of the solution vectors should be private content of the time integrator in my opinion.

My understanding of a function like `get_solution_np()` is to receive the solution at time `t_n+1` as long as the time step `n -> n + 1` has not been completed. Once a time step is completed (but a new one has not been initiated so far), the function `get_solution_np()` does not have a meaning, because we do not have a guarantee that `solution_np` exists at all. A time integrator could realize `solution_np` as a temporary vector which is deleted once a time step has been finished. This could be the reason why there is currently no `get_solution_np()` functionality in general.

Reply:
this one is still open, will look at it tomorrow morning! :+1:

Reply:
I agree on the interpretation of function `time_integrator->get_solution_np()`.
I think the naming I introduced in this PR is very misleading, since there is a member variable `solution_np`, but we actually return the most recent solution vector.

The `solution_np` vector contains the extrapolation at some point in the time stepper, but I wanted to have a solution at the current or even previous timestep. The sole purpose was to have some (close to?) consistent and recent enough vector for AMR.

In AMR, we call what is now named `time_integrator->get_solution_np()` after `time_integrator->advance_one_timestep_solve()`, but before updating the vectors in `time_integrator->advance_one_timestep_post_solve()`. So, the vectors are actually not the same, with `solution[0]` containing the old time step's solution still, but `solution_np` having the most recent solution (which has just been computed at this point).

Long story short: I will just return `solution_np` as the name suggests, ok?
 
Interestingly, the error increases now a tiny bit for the tests, but the mesh is also different.
So I think its just in general better to use the most recent vector `solution_np`.

Reply:
see
https://github.com/exadg/exadg/pull/586/commits/62f80c3b6cda67b715b8b3099a3bc18532f6b2b4